graphblas_algorithms/algorithms/shortest_paths/weighted.py

```diff
@@ -164,6 +165,117 @@ def bellman_ford_path_lengths(G, nodes=None, *, expand_output=False):
+    d = dict(zip(indices.tolist(), values.tolist()))
+    if dst not in d:
+        return []
+    cur = dst
+    path = [cur]
+    while cur != src:
+        cur = d[cur]
+        path.append(cur)
+    return G.list_to_keys(reversed(path))
+
+
+def bellman_ford_path(G, source, target):
+    src_id = G._key_to_id[source]
+    dst_id = G._key_to_id[target]
+    if G.get_property("is_iso"):
+        # If the edges are iso-valued (and positive), then we can simply do level BFS
+        is_negative = G.get_property("has_negative_edges+")
+        if not is_negative:
+            p = _bfs_parent(G, source, target=target)
+            return _reconstruct_path_from_parents(G, p, src_id, dst_id)
+        raise Unbounded("Negative cycle detected.")
+    A, is_negative, has_negative_diagonal = G.get_properties(
+        "offdiag has_negative_edges- has_negative_diagonal"
+    )
+    if A.dtype == bool:
+        # Should we upcast e.g. INT8 to INT64 as well?
+        dtype = int
+    else:
+        dtype = A.dtype
+    cutoff = None
+    n = A.nrows
+    d = Vector(dtype, n, name="bellman_ford_path_length")
+    d[src_id] = 0
+    p = Vector(int, n, name="bellman_ford_path_parent")
+    p[src_id] = src_id
+
+    prev = d.dup(name="prev")
+    cur = Vector(dtype, n, name="cur")
+    indices = Vector(int, n, name="indices")
+    mask = Vector(bool, n, name="mask")
+    B = Matrix(dtype, n, n, name="B")
+    Indices = Matrix(int, n, n, name="Indices")
+    cols = prev.to_coo(values=False)[0]
+    one = unary.one[bool]
+    for _i in range(n - 1):
+        # This is a slightly modified Bellman-Ford algorithm.
+        # `cur` is the current frontier of values that improved in the previous iteration.
+        # This means that in this iteration we drop values from `cur` that are not better.
+        cur << min_plus(prev @ A)
+        if cutoff is not None:
+            cur << select.valuele(cur, cutoff)
+
+        # Mask is True where cur not in d or cur < d
+        mask << one(cur)
+        mask(binary.second) << binary.lt(cur & d)
+
+        # Drop values from `cur` that didn't improve
+        cur(mask.V, replace) << cur
+        if cur.nvals == 0:
+            break
+        # Update `d` with values that improved
+        d(cur.S) << cur
+        if not is_negative:
+            # Limit exploration if we have a target
+            cutoff = cur.get(dst_id, cutoff)
+
+        # Now try to find the parents!
+        # This is also not standard. Typically, UDTs and UDFs are used to keep
+        # track of both the minimum element and the parent id at the same time.
+        # Only include rows and columns that were used this iteration.
+        rows = cols
+        cols = cur.to_coo(values=False)[0]
+        B.clear()
+        B[rows, cols] = A[rows, cols]
+
+        # Reverse engineer to determine parent
+        B << binary.plus(prev & B)
+        B << binary.iseq(B & cur)
+        B << select.valuene(B, False)
+        Indices << indexunary.rowindex(B)
```

Comment:
These two lines could be replaced with
```python
Indices(B.V, replace) << indexunary.rowindex(B)
```
but I think it's clearer in two lines (using masking with replace is kind of ugly). I don't know if there is a performance difference.